### PR TITLE
Report no AssemblyReferences from corlib in EE

### LINKED ISSUE #13275: EE: Implement ProjectN corlib workaround
The current ProjectN corlib has AssemblyRefs and thus isn't considered a valid corlib by the compiler. We need to implement a workaround for EE to enable compilation against such corlib. Long term the CLR is working towards removing the AssemblyRefs.

### PR DESCRIPTION
**Pull request by cston**
Add an `ignoreAssemblyRefs` parameter to `ModuleMetadata` to allow the EE to expose .NET Native corlib as an assembly with no references.

Fixes #13275.

### CODE REVIEW

**Comment by cston** [comment]
@tmat, @jaredpar, @gafter, please review. Will add tests for C# and VB if this approach seems appropriate.

[Thanks @tmat for suggesting this work around.]

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 1]
```diff
@@ -42,11 +41,31 @@ internal static class MetadataUtilities
...
             {
                 var metadata = ModuleMetadata.CreateFromMetadata(block.Pointer, block.Size, includeEmbeddedInteropTypes: true);
                 try
                 {
+                    var reader = metadata.MetadataReader;
+                    if (corLibrary == null)
+                    {
+                        bool hasNoAssemblyRefs = reader.AssemblyReferences.Count == 0;
+                        // .NET Native uses a corlib with references.
+                        if (hasNoAssemblyRefs || metadata.Name.Equals("System.Private.CoreLib.dll", StringComparison.OrdinalIgnoreCase))
```
Comment (tmat) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Could you add a link to an issue to remove this workaround?

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ReferencedModulesTests.cs [hunk 2]
```diff
@@ -705,6 +706,55 @@ .maxstack  2
...
+            var source =
+@"class C
+{
+    static void M()
+    {
+    }
+}";
+            unsafe
+            {
+                var bytes = ImmutableArray<byte>.Empty; // TODO
```
Comment (tmat) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ReferencedModulesTests.cs]: Assuming the TODO is to replace with actual bytes...

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs [hunk 3]
```diff
@@ -84,7 +83,7 @@ internal sealed class PEModule : IDisposable
         private static readonly AttributeValueExtractor<ObsoleteAttributeData> s_attributeObsoleteDataExtractor = CrackObsoleteAttributeData;
         private static readonly AttributeValueExtractor<ObsoleteAttributeData> s_attributeDeprecatedDataExtractor = CrackDeprecatedAttributeData;
 
-        internal PEModule(ModuleMetadata owner, PEReader peReader, IntPtr metadataOpt, int metadataSizeOpt, bool includeEmbeddedInteropTypes = false)
+        internal PEModule(ModuleMetadata owner, PEReader peReader, IntPtr metadataOpt, int metadataSizeOpt, bool includeEmbeddedInteropTypes, bool ignoreAssemblyRefs)
```
Comment (tmat) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: Comment and link to an issue to remove.

**Comment by tmat** [comment]
👍

**Comment by cston** [comment]
Adding @dotnet/roslyn-compiler, @tommcdon.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs [hunk 4]
```diff
@@ -42,11 +41,31 @@ internal static class MetadataUtilities
...
+                        {
+                            // If this assembly declares System.Object, assume it is the corlib.
+                            if (reader.DeclaresTheObjectClass())
+                            {
+                                corLibrary = reader.ReadAssemblyIdentityOrThrow();
```
Comment (AlekseyTs) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: >corLibrary = reader.ReadAssemblyIdentityOrThrow(); [](start = 32, length = 50)

It looks like the value stored in ```corLibrary``` is order dependent. Is this a concern?

**Comment by AlekseyTs** [comment]
LGTM. Please add comments and references to issues requested by @tmat.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: Added comment and link to the original issue.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/MetadataDecoder.cs [hunk 5]
```diff
@@ -127,8 +127,16 @@ protected override TypeSymbol LookupNestedTypeDefSymbol(TypeSymbol container, re
...
-                return assembly.LookupTopLevelMetadataType(ref emittedName, digThroughForwardedTypes: true);
+                var referencedAssemblies = moduleSymbol.GetReferencedAssemblySymbols();
+                if (referencedAssemblyIndex < referencedAssemblies.Length)
+                {
+                    AssemblySymbol assembly = referencedAssemblies[referencedAssemblyIndex];
+                    return assembly.LookupTopLevelMetadataType(ref emittedName, digThroughForwardedTypes: true);
+                }
+                else
+                {
+                    return new UnsupportedMetadataTypeSymbol();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/MetadataDecoder.cs]: >return new UnsupportedMetadataTypeSymbol(); [](start = 20, length = 43)

Please add a comment about when do we expect to get here. Also, please add an assert or better throw if current assembly is not a core assembly.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs [hunk 6]
```diff
@@ -651,38 +651,44 @@ internal NamedTypeSymbol LookupTopLevelMetadataType(ref MetadataTypeName emitted
...
+            catch (BadImageFormatException)
+            {
+                return null;
+            }
+            var referencedAssemblies = this.GetReferencedAssemblySymbols();
+            if (index < referencedAssemblies.Length)
+            {
+                return referencedAssemblies[index];
+            }
+            return null;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs]: >return null; [](start = 12, length = 12)

Same suggestion as in MetadataDecoder.cs

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs [hunk 7]
```diff
@@ -651,38 +651,44 @@ internal NamedTypeSymbol LookupTopLevelMetadataType(ref MetadataTypeName emitted
...
                 var name = MetadataTypeName.FromFullName(forwarder.Key);
-                AssemblySymbol assemblySymbol;
-
-                try
-                {
-                    assemblySymbol = this.GetReferencedAssemblySymbols()[Module.GetAssemblyReferenceIndexOrThrow(forwarder.Value)];
-                }
-                catch (BadImageFormatException)
+                var assemblySymbol = this.GetReferencedAssemblySymbol(forwarder.Value);
+                if (assemblySymbol == null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs]: >assemblySymbol == null [](start = 20, length = 22)

(object)?

**Comment by AlekseyTs** [comment]
It feels like similar changes should be made to VB compiler.

**Comment by cston** [comment]
@AlekseyTs, ported changes to VB.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs (hunk 7 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEModuleSymbol.cs]: Fixed.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/MetadataDecoder.cs (hunk 5 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/MetadataDecoder.cs]: Added comment and throw to new `GetReferencedAssemblySymbol` method.

File: src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs (hunk 4 above)
Comment (cston) [inline src/ExpressionEvaluator/Core/Source/ExpressionCompiler/MetadataUtilities.cs]: Yes, we'll choose the first assembly that meets the corlib requirements, even if there are multiple. That should be acceptable for evaluating expressions in the EE. Added a comment indicating that.

File: src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEModuleSymbol.vb [hunk 8]
```diff
@@ -435,28 +435,33 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
...
                     Continue For
-                End Try
+                End If
 
                 Dim name = MetadataTypeName.FromFullName(forwarder.Key)
                 Yield assembly.LookupTopLevelMetadataType(name, digThroughForwardedTypes:=True)
             Next
         End Function
 
+        Private Shadows Function GetReferencedAssemblySymbol(assemblyRef As AssemblyReferenceHandle) As AssemblySymbol
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEModuleSymbol.vb]: >Shadows [](start = 16, length = 7)

Overloads?

File: src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEModuleSymbol.vb [hunk 9]
```diff
@@ -435,28 +435,33 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
...
+            Dim referencedAssemblyIndex As Integer
+            Try
+                referencedAssemblyIndex = Me.Module.GetAssemblyReferenceIndexOrThrow(assemblyRef)
+            Catch ex As BadImageFormatException
+                Return Nothing
+            End Try
+            Dim result = MyBase.GetReferencedAssemblySymbol(referencedAssemblyIndex)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEModuleSymbol.vb]: >MyBase [](start = 25, length = 6)

I think ```MyBase``` wouldn't be required if this method was Overloads.

File: src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEModuleSymbol.vb [hunk 10]
```diff
@@ -435,28 +435,33 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
...
+            Debug.Assert(result IsNot Nothing) ' TODO: Test missing assembly case!
```
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEModuleSymbol.vb]: Yes, I am investigating testing these cases.

File: src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/VisualBasicEESymbolProvider.vb [hunk 11]
```diff
@@ -66,7 +66,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator
         ''' <exception cref="BadImageFormatException"></exception>
         Public Overrides Function GetReferencedAssembly(handle As AssemblyReferenceHandle) As IAssemblySymbol
             Dim index As Integer = _metadataDecoder.Module.GetAssemblyReferenceIndexOrThrow(handle)
-            Return _metadataDecoder.ModuleSymbol.GetReferencedAssemblySymbols()(index)
+            Dim assembly = _metadataDecoder.ModuleSymbol.GetReferencedAssemblySymbol(index)
+            If assembly Is Nothing Then
+                Debug.Assert(False) ' TODO: Test missing assembly case!
```
Comment (AlekseyTs) [inline src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/VisualBasicEESymbolProvider.vb]: >Debug.Assert(False) ' TODO: Test missing assembly case! [](start = 16, length = 55)

Is this temporary code?

File: src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEModuleSymbol.vb (hunk 9 above)
Comment (cston) [inline src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEModuleSymbol.vb]: Fixed.

**Comment by cston** [comment]
@tmat, @jaredpar, @alekseyts, please review the latest changes - there were several callers of `PEModuleSymbol.GetReferenceAssemblySymbols()` that had to be updated to handle assembly index out of range when testing references to missing assemblies. Thanks.

**Comment by cston** [comment]
@dotnet-bot retest windows_debug_unit64_prtest please

**Comment by cston** [comment]
**Customer scenario:** Moving remote debugging to portable CoreCLR.

**Bug:** https://github.com/dotnet/roslyn/issues/13275

**Workarounds:** None

**Risk:** The EE reports no assembly references from the .NET Native corlib, although the corlib metadata still contains type references to those missing assemblies. Therefore, the few code paths that handle assembly references now handle missing assemblies. If there are other code paths we've overlooked, the EE will fail to evaluate expressions in those cases when the expression or the current frame refers to types from the missing assemblies.

**Performance impact:** The EE was already determining which assembly was the corlib, so the overhead of recognizing this specific corlib should be minimal.

**Regression?** No

**How was bug found?** Found when adding EE support for .NET Native.

**Comment by cston** [comment]
@mattgertz for approval.

**Comment by MattGertz** [comment]
@AlekseyTs All good?

**Comment by cston** [comment]
@dotnet-bot retest windows_debug_unit32_prtest please

**Comment by cston** [comment]
@dotnet-bot retest this please

**Pull request merged**